cmd/ecc/join.go

```diff
@@ -0,0 +1,330 @@
+// 1) Performing as few read operations as possible overall
+//    while doing as many reads in parallel as possible.
+// 2) Reading data shards first so that no erasure reconstruction
+//    is necessary if all data shards are available.
+//
+// Before changing the implementation make sure that you understand
+// the implications and take a look at the approaches [1] and [2]
+// at the end of the file - which are less optimal.
+
+const (
+	// ErrReadQuorum is the error returned by the JoinedReaders
+	// when too many read operations fail such that the erasure
+	// coding won't be able to reconstruct the content data.
+	ErrReadQuorum errorType = "ecc: no read quorum: too many data sources are unavailable"
+
+	// errOffline is the error used to mark a data source as offline.
+	// The MinIO server may has failed to fetch the metadata (XL.json)
+	// from certain data sources before. Such data sources are marked as
+	// offline by an io.ReaderAt that is nil.
+	errOffline errorType = "ecc: dat source is marked as offline"
+)
+
+// JoinReaders combines a set of data sources. It implements reading
+// from the sources concurrently. During a read it spawns and waits
+// for go routines that read from data sources until the all read
+// operations have accumulated enough data to reconstruct it.
+func JoinReaders(src []io.ReaderAt, offset int64) *JoinedReaders {
+	errors := make([]error, len(src))
+	for i := range src {
+		if src[i] == nil { // If a data source is nil...
+			errors[i] = errOffline // ...mark it as offline.
+		}
+	}
+	return &JoinedReaders{
+		src:    src,
+		err:    errors,
+		offset: offset,
+	}
+}
+
+// JoinedReaders groups a set of data sources
+// and manages reading from them concurrently.
+type JoinedReaders struct {
+	src    []io.ReaderAt
+	err    []error
+	offset int64
+
+	rErr error
+}
+
+// Read tries to read just as many shards
+// as necessary into the buffer such that the
+// erasure coding can reconstruct missing data
+// shards, if any.
+//
+// Let d be the number of data shards. Read spawns
+// d go routines - each tries to read one complete
+// shard from its data source. Whenever a go routine
+// completes but returns an error this data source
+// is considered unavailable and another go routine
+// is spawned. This new go routine tries to read from
+// the next data source. If no more data sources are
+// available it returns ErrNoReadQuorum.
+//
+// Read minimizes the total number of reads / connections
+// and tries to read as many data shards before
+// it tries to fetch any parity shards. Minimizing
+// the number of reads / connections is important to
+// avoid any unnecessary I/O or network load. Also, if
+// a data shard is available we should prefer fetching it
+// instead of reading a parity shard to avoid unnecessary
+// erasure coding reconstruction.
+func (r *JoinedReaders) Read(b *Buffer) error {
+	if len(r.src) != len(b.shards) {
+		panic("ecc: number of data sources does not match the number of shards")
```

Comment:
This is a programmer mistake exposing a logical mistake made by the caller. It indicates that the calling code contains a bug and the only way to recover from this "error" is: the programmer fixes the code.

Reply:
I like this answer by Ian Lance Taylor:

> Ian Lance Taylor, Free software programmer, Go developer, Google employee

> You don't ever need to use panic.

> It's reasonable to use panic when a situation that "can't happen" does in fact happen.  That is, call panic when there is some logic error in your program.

> It's also reasonable to use panic within a single package to simplify the code, with calls to recover within that package to prevent to panic from leaking out.  For example, the standard encoding/gob package does this.  It's careful to use a package-local type to distinguish panics that should be caught from panics that were triggered by code called by the package.

And from "Effective Go": 

> [...] but real library functions should avoid panic. If the problem can be masked or worked around, it's always better to let things continue to run rather than taking down the whole program. One possible counterexample is during initialization: if the library truly cannot set itself up, it might be reasonable to panic, so to speak.

Since this an exported function you should not panic IMO.